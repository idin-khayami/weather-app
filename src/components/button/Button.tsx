import { ReactNode } from 'react';

import clsx from 'clsx';

import { Loading } from '@/components/loading';

import styles from './button.module.scss';

interface ButtonProps {
  children: ReactNode;
  onClick?: () => void;
  isLoading?: boolean;
  className?: string;
}

export const Button = ({
  children,
  onClick,
  className,
  isLoading = false,
}: ButtonProps) => {
  return (
    <button
      className={clsx(styles.buttonBase, className)}
      onClick={onClick}
      disabled={isLoading}
    >
      {isLoading ? <Loading /> : children}
    </button>
  );
};
