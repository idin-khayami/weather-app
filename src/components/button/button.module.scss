@import '../../styles/vars';

.buttonBase {
  border: none;
  display: flex;
  min-width: 50px;
  min-height: 50px;
  justify-content: center;
  align-items: center;
  padding: $buttonPadding;
  border-radius: 10px;
  font-size: $large;
  font-weight: 900;
  background: $buttonGradient;
  color: $buttonForeground;
  aspect-ratio: 1; /* Ensures the button remains square */
  &:hover {
    cursor: pointer;
  }
}
