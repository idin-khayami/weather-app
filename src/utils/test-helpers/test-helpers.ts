import { ReactElement } from 'react';

import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { UserEvent, userEvent } from '@testing-library/user-event';

export const renderWithUserSetup = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult & { user: UserEvent } => {
  return {
    user: userEvent.setup(),
    ...render(ui, options),
  };
};
