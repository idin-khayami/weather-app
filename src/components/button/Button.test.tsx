import { render, screen } from '@testing-library/react';

import { Button } from '@/components/button/Button';
import { renderWithUserSetup } from '@/utils/test-helpers/test-helpers';

describe('Component: button component', () => {
  it('should render', () => {
    render(<Button>This is a Button</Button>);

    const button = screen.getByRole('button');

    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent('This is a Button');
  });

  it('should match the snapshot', () => {
    render(<Button>This is a Button</Button>);

    const button = screen.getByRole('button');

    expect(button).toMatchSnapshot();
  });

  it('should be clickable', async () => {
    const mockFn = jest.fn();

    const { user } = renderWithUserSetup(
      <Button onClick={mockFn}>This is a Button</Button>,
    );

    const button = screen.getByRole('button');

    expect(mockFn).toHaveBeenCalledTimes(0);
    await user.click(button);
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('should be disabled when loading', async () => {
    const mockFn = jest.fn();

    const { user } = renderWithUserSetup(
      <Button onClick={mockFn} isLoading={true}>
        This is a Button
      </Button>,
    );

    const button = screen.getByRole('button');

    expect(mockFn).toHaveBeenCalledTimes(0);
    await user.click(button);
    expect(mockFn).toHaveBeenCalledTimes(0);
  });

  it('sample functionality test', async () => {
    let variable = 0;

    const { user } = renderWithUserSetup(
      <Button
        onClick={() => {
          variable = variable + 1;
        }}
      >
        This is a Button
      </Button>,
    );

    const button = screen.getByRole('button');

    expect(variable).toBe(0);
    await user.click(button);
    expect(variable).toBe(1);
  });
});
