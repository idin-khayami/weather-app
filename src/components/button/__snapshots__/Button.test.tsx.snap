// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: button component should match the snapshot 1`] = `
<button
  class="buttonBase primary"
>
  This is a Button
</button>
`;
